gensim/models/doc2vec.py

```diff
@@ -181,68 +496,175 @@ def __init__(self, sentences=None, size=300, alpha=0.025, window=8, min_count=5,
         specifies how many "noise words" should be drawn (usually between 5-20).
 
         `dm_mean` = if 0 (default), use the sum of the context word vectors. If 1, use the mean.
-        Only applies when dm is used.
+        Only applies when dm is used in non-concatenative mode.
+
+        `dm_concat` = if 1, use concatenation of context vectors rather than sum/average;
+        default is 0 (off). Note concatenation results in a much-larger model, as the input
+        is no longer the size of one (sampled or arithmatically combined) word vector, but the
+        size of the tag(s) and all words in the context strung together.
+
+        `dm_tag_count` = expected constant number of document tags per document, when using
+        dm_concat mode; default is 1.
+
+        `dbow_words` if set to 1 trains word-vectors (in skip-gram fashion) simultaneous with DBOW
+        doc-vector training; default is 0 (faster training of doc-vectors only).
 
         """
         Word2Vec.__init__(self, size=size, alpha=alpha, window=window, min_count=min_count,
                           sample=sample, seed=seed, workers=workers, min_alpha=min_alpha,
-                          sg=(1+dm) % 2, hs=hs, negative=negative, cbow_mean=dm_mean, **kwargs)
-        self.train_words = train_words
-        self.train_lbls = train_lbls
-        if sentences is not None:
-            self.build_vocab(sentences)
-            self.train(sentences)
-
-    @staticmethod
-    def _vocab_from(sentences):
-        sentence_no, vocab = -1, {}
+                          sg=(1+dm) % 2, hs=hs, negative=negative, cbow_mean=dm_mean, 
+                          null_word=dm_concat, **kwargs)
+        self.dbow_words = dbow_words
+        self.dm_concat = dm_concat
+        self.dm_tag_count = dm_tag_count
+        self.docvecs = docvecs
+        if not self.docvecs:
+            self.docvecs = DocvecsArray(docvecs_mapfile)
+        self.comment = comment
+        if documents is not None:
+            self.build_vocab(documents)
+            self.train(documents)
+
+    def clear_sims(self):
+        Word2Vec.clear_sims(self)
```

Comment:
Ok. Note that this pattern exists in a few other places, e.g., the next following couple of methods.